### ISSUE TITLE
Thumbnailer: icons have wrong/mixed size on HiDPI screens

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#783]
Icons have wrong/mixed size on HiDPI screens. Example: Ubuntu 18.04 on a HiDPI machine.

![](https://user-images.githubusercontent.com/2480569/40280879-1f70e55a-5c49-11e8-894c-be7db6b9dc5d.png)

**Comment by TheAssassin** [comment]
Did you check whether there's big icons at all in these AppImages? It seems to me like the PdftoWord AppImage's icon has a higher resolution than all the other ones'.

**Comment by probonopd** [comment]
May be the case, nevertheless it looks like the thumbnailer needs to upscale the lower-resolution icons...

**Comment by azubieta** [comment]
This issue doesn't happen with the kde thumbnailer because it does chamges
the images size to match the standard.
I've investigating a bit and I think that thunar is capable of using gnome
thumbnailers. So if we make a gnome thumbailer we will have covered the
major DE. I'm just wondering if there is a simple thumbailer code that I
could use as template.

**Comment by probonopd** [comment]
> linking to some library like imagemagick

Possibly https://lazka.github.io/pgi-docs/GdkPixbuf-2.0/classes/Pixbuf.html#GdkPixbuf.Pixbuf.scale_simple

**Comment by probonopd** [comment]
Ideally, the desktop environment would do the resizing, but apparently not all do. So we need to do the resizing in the thumbnailer, don't we?

**Comment by azubieta** [comment]
We could do it at the moment of packing by means of "linuxdeploy" so the two icons get created. There is no problem of including image manipulation libraries.
An yes we should war developers about missing icons at a given resolution.

**Comment by probonopd** [comment]
> We could do it at the moment of packing by means of "linuxdeploy" so the two icons get created. There is no problem of including image manipulation libraries.

Yes, or even in the lower-level `appimagetool` that gets used by most AppImage creation methods.

**Comment by TheAssassin** [comment]
appimagetool shouldn't take care of icon manipulation. It should only package what it gets presented.

**Comment by probonopd** [comment]
Using the UNIX philosophy, we should have a `makeicons` command line tool that would put the appropriate icons in the appropriate sizes into the appropriate locations in a user-defined `$PREFIX` a.k.a. AppDir structure.

**Comment by TheAssassin** [comment]
I'd rather build something into linuxdeploy which upscales icons, and prints warnings about the lack of HiDPI support.

I am strongly opposed to appimagetool itself containing functionality regarding desktop integration. This is out of scope.

**Comment by probonopd** [comment]
Whatever we put in linuxdeploy will not end up being used by
* electron-builder
* Handmade/scripted AppImages
* linuxdeployqt (might go away _one day_)

**Comment by TheAssassin** [comment]
@probonopd that was just an example. The point is, adding image processing to AppImageKit adds a _lot_ of dependencies. And since this is desktop integration stuff, it's actually out of scope. AppImageKit change too much in a given AppDir to keep it simple to deploy.

If you think there is the need for an external solution, we should really think about creating a separate tool/library. Heck, we could simply write it as a linuxdeploy plugin, as they're standalone tools, they can be used on _any_ AppDir, independently from linuxdeploy.

**Comment by probonopd** [comment]
Architecturally, `appimagetool` currently takes an _existing_ AppDir and puts it into an AppImage _unchanged_. (My test for this is that I frequently mount an AppImage and run `appimagetool` on the read-only mount point of the mounted AppImage - it will produce a new AppImage). Possibly it should stay like this.

So I think we should have a command line tool that would do the modification/preparation/checking of the AppDir. Possibly the tool scope of that tool could even be more than just icons - e.g., linting the AppDir (doing plausibility checks), taking care of metadata, AppStream information, etc. - everything that makes a "good AppDir"?

But ideally not baked into `linuxdeployqt`, `linuxdeploy`, etc. because a) there's always more than 1 such tool and b) not everyone will use those tools.

**Comment by TheAssassin** [comment]
linuxdeploy provides a standalone `AppDir` class, which will soon receive a `verify()` method that will implement exactly what you said.

**Comment by probonopd** [comment]
Which is probably shorthand for "the standalone command line tool that takes care of the icons could also be used as a `linuxdeploy`plugin?

**Comment by TheAssassin** [comment]
No. All I'm saying is that liblinuxdeploy_core will receive that feature. You could use it to build your standalone tool if you want to. You know that I build my software so that it's as reusable as possible.

**Comment by probonopd** [comment]
Why not make a separate plugin = standalone executable for the icon handling stuff? Then other consumers (outside of `linuxdeploy`) could just invoke it without having to "extract it" from `liblinuxdeploy_core`...

**Comment by TheAssassin** [comment]
The `verify` function is just a linter, it doesn't modify the deployed icons. You mentioned you wanted hints that e.g., icons are too small, and that's exactly what a verification function should recognize.

The problem with an icon is that at the moment, the linuxdeploy plugin spec requires plugins to not depend on any kind of order. They must be able to be run in any order. We'd have to introduce plugin priorities or anything like that to make sure such a plugin would be the last one that is run. That's not really trivial.

**Comment by probonopd** [comment]
Or we make the plugin not just lint the icons, but create/fix them, right?

**Comment by TheAssassin** [comment]
That'd be the plugin's scope, yes. AppDir's verify() shall just check for common issues and tell the user about them, but won't yield errors and interrupt linuxdeploy's workflow.

All I'm saying is that if it'll be a plugin for linuxdeploy, we need to find a way to make sure it's the last thing to be run unless you think that won't be a problem.
